File: js/mandelbrot.js

// Ref: http://progur.com/2017/02/create-mandelbrot-fractal-javascript.html

(function() {

    var myCanvas = document.createElement("canvas");
    myCanvas.width = 600;
    myCanvas.height = 600;
    document.body.appendChild(myCanvas);
    var ctx = myCanvas.getContext("2d");

    var maxIterations = 100;
    var magnificationFactor = 300;  // 2900
    var panX = 1.5;  // 0.6
    var panY = 1;  // 0.7

    function checkIfBelongsToMandelbrotSet(x, y) {
        var realComponentOfResult = x;
        var imaginaryComponentOfResult = y;
        for (var i = 0; i < maxIterations; i++) {
            var tempRealComponent = realComponentOfResult * realComponentOfResult
                - imaginaryComponentOfResult * imaginaryComponentOfResult
                + x;
            var tempImaginaryComponent = 2 * realComponentOfResult * imaginaryComponentOfResult
                + y;
            realComponentOfResult = tempRealComponent;
            imaginaryComponentOfResult = tempImaginaryComponent;

            // Return a number as a percentage
            if (realComponentOfResult * imaginaryComponentOfResult > 5) {
                return (i / maxIterations * 100);
            }
        }
        return 0;   // Return zero if in set
    }

    for (var x = 0; x < myCanvas.width; x++) {
        for (var y = 0; y < myCanvas.height; y++) {
            var belongsToSet =
                checkIfBelongsToMandelbrotSet(
                    x / magnificationFactor - panX,
                    y / magnificationFactor - panY
                );
            if (belongsToSet) {
                if (belongsToSet === 0) {
                    ctx.fillStyle = '#000';
                    ctx.fillRect(x,y, 1,1); // Draw a black pixel
                } else {
                    ctx.fillStyle = 'hsl(0, 100%, ' + belongsToSet + '%)';
                    ctx.fillRect(x,y, 1,1); // Draw a colorful pixel
                }
            }
        }
    }

})();

File: src/mandelbrot.rs

// Functions that you wish to access from Javascript must be marked as no_mangle
#[no_mangle]
pub fn belongs_to_mandelbrot_set(x: f64, y: f64, max_iterations: u64) -> f64 {
    // println!("{}", format!("!!! {} {} {}", x, y, max_iterations));

    let mut real_component_of_result: f64 = x;
    let mut imaginary_component_of_result: f64 = y;

    for i in 0..max_iterations {
        let temp_real_component: f64 = real_component_of_result * real_component_of_result
            - imaginary_component_of_result * imaginary_component_of_result
            + x;
        let temp_imaginary_component: f64 = 2.0 * real_component_of_result * imaginary_component_of_result
            + y;
        real_component_of_result = temp_real_component;
        imaginary_component_of_result = temp_imaginary_component;

        // Color: return a percentage
        if real_component_of_result * imaginary_component_of_result > 5.0 {
            return i as f64 / max_iterations as f64 * 100.0;
        }
    }
    return 0.0;   // Return zero if in set
}

fn main() {
    println!("mais.rs: My fractal application in Rust");
}

